# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1731609295 -3600
# Parent  31ca2df3290c4a8c74ebb9dba3bce141e5cd2c05
Implement negation for QuickFilterBar filter and text filter buttons, prevent clicks while previous click hasn't been processed.

diff --git a/mail/modules/QuickFilterManager.sys.mjs b/mail/modules/QuickFilterManager.sys.mjs
--- a/mail/modules/QuickFilterManager.sys.mjs
+++ b/mail/modules/QuickFilterManager.sys.mjs
@@ -1037,25 +1037,29 @@ var TagFacetingFilter = {
     } else {
       aState.mode = qbm.value;
     }
 
-    // When the tag filter is inverted, we only allow either selection or
-    // inversion of tag buttons.
-    // If we allowed both, the backend can't correctly process the resulting
-    // expression.
-    let tagsSelected = 0;
-    let tagsInverted = 0;
-    for (const tagKey in keywordMap) {
-      const value = keywordMap[tagKey];
-      if (value === true) {
-        tagsSelected++;
-      } else if (value === false) {
-        tagsInverted++;
+    function countTags() {
+      // When the tag filter is inverted, we only allow either selection or
+      // inversion of tag buttons.
+      // If we allowed both, the backend can't correctly process the resulting
+      // expression.
+      let tagsSelected = 0;
+      let tagsInverted = 0;
+      for (const tagKey in keywordMap) {
+        const value = keywordMap[tagKey];
+        if (value === true) {
+          tagsSelected++;
+        } else if (value === false) {
+          tagsInverted++;
+        }
       }
+      qbm.disabled = !tagsSelected && !tagsInverted;
+      return { tagsSelected, tagsInverted };
     }
-    qbm.disabled = !tagsSelected && !tagsInverted;
 
     function clickHandler() {
+      const { tagsInverted } = countTags();
       const tagKey = this.getAttribute("value");
       if (keywordMap[tagKey] === null && aState.inverted && tagsInverted) {
         // We use a delay to clear the button to give feedback that an action
         // was indeed performed.
@@ -1068,8 +1072,9 @@ var TagFacetingFilter = {
       aMuxer.deferredUpdateSearch();
     }
 
     function rightClickHandler(aEvent) {
+      const { tagsSelected } = countTags();
       if (aEvent.button == 2) {
         // A |toggle-button| sets up pressed and inverted in the constructor.
         const tagKey = this.getAttribute("value");
         const tagKeyState = keywordMap[tagKey];
